File: 3d-app-components/layers/layer3.js
// Import necessary libraries
import * as THREE from 'three';

// Function to compute FFT and detect phase inversion
function spectralPhaseInversion(prices, windowSize = 64) {
    const fftResults = [];
    const phaseInversions = [];
    
    for (let i = 0; i < prices.length - windowSize; i++) {
        const window = prices.slice(i, i + windowSize);
        const fftResult = new FFT(windowSize).forward(window);
        const dominantFreq = fftResult.map((_, index) => index * (1 / windowSize)).slice(0, windowSize / 2);
        const dominantAmplitude = fftResult.map(Math.abs).slice(0, windowSize / 2);
        
        // Detect dominant frequency
        const fStar = dominantFreq[dominantAmplitude.indexOf(Math.max(...dominantAmplitude))];
        const phaseAngle = Math.atan2(fftResult[dominantAmplitude.indexOf(Math.max(...dominantAmplitude))].imag, 
                                         fftResult[dominantAmplitude.indexOf(Math.max(...dominantAmplitude))].real);
        
        // Check for phase inversion condition
        if (Math.abs(phaseAngle) > Math.PI / 2) {
            phaseInversions.push(i + Math.floor(windowSize / 2));
        }
        
        fftResults.push({ fStar, phaseAngle });
    }
    
    return { fftResults, phaseInversions };
}

// Sample price data (replace with actual price data)
const prices = Array.from({ length: 1000 }, () => Math.random());  // Simulated price data

// Run the spectral phase inversion detection
const { fftResults, phaseInversions } = spectralPhaseInversion(prices);

// Visualization using three.js
const scene = new THREE.Scene();
const camera = new THREE.PerspectiveCamera(75, window.innerWidth / window.innerHeight, 0.1, 1000);
const renderer = new THREE.WebGLRenderer();
renderer.setSize(window.innerWidth, window.innerHeight);
document.body.appendChild(renderer.domElement);

// Create a line for price data
const priceGeometry = new THREE.BufferGeometry();
const priceVertices = new Float32Array(prices);
priceGeometry.setAttribute('position', new THREE.BufferAttribute(priceVertices, 3));
const priceMaterial = new THREE.LineBasicMaterial({ color: 0x0000ff });
const priceLine = new THREE.Line(priceGeometry, priceMaterial);
scene.add(priceLine);

// Add phase inversion markers
phaseInversions.forEach(inversion => {
    const markerGeometry = new THREE.SphereGeometry(0.1, 32, 32);
    const markerMaterial = new THREE.MeshBasicMaterial({ color: 0xff0000 });
    const marker = new THREE.Mesh(markerGeometry, markerMaterial);
    marker.position.set(inversion, prices[inversion], 0);
    scene.add(marker);
});

// Set camera position
camera.position.z = 5;

// Animation loop
function animate() {
    requestAnimationFrame(animate);
    renderer.render(scene, camera);
}
animate();

File: 3d-app-components/layers/layer3.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Function to compute FFT and detect phase inversion
def spectral_phase_inversion(prices, window_size=64):
    fft_results = []
    phase_inversions = []
    
    for i in range(len(prices) - window_size):
        window = prices[i:i + window_size]
        fft_result = np.fft.fft(window)
        dominant_freq = np.fft.fftfreq(len(window), d=1)[:len(window)//2]
        dominant_amplitude = np.abs(fft_result)[:len(window)//2]
        
        # Detect dominant frequency
        f_star = dominant_freq[np.argmax(dominant_amplitude)]
        phase_angle = np.angle(fft_result[np.argmax(dominant_amplitude)])
        
        # Check for phase inversion condition
        if abs(phase_angle) > np.pi / 2:
            phase_inversions.append(i + window_size // 2)
        
        fft_results.append((f_star, phase_angle))
    
    return fft_results, phase_inversions

# Sample price data (replace with actual price data)
prices = np.random.rand(1000)  # Simulated price data

# Run the spectral phase inversion detection
fft_results, phase_inversions = spectral_phase_inversion(prices)

# Visualization
plt.figure(figsize=(14, 7))
plt.plot(prices, label='Price Data')
plt.title('Price Data with Spectral Phase Inversion Detection')
plt.xlabel('Time')
plt.ylabel('Price')
plt.axvline(x=phase_inversions[0], color='red', linestyle='--', label='Phase Inversion')
for inversion in phase_inversions:
    plt.axvline(x=inversion, color='red', linestyle='--')
plt.legend()
plt.show()
